profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LibrariesInfoCache.cpp

```diff
@@ -0,0 +1,111 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+
+#include "LibrariesInfoCache.h"
+
+#include "Log.h"
+
+#include <string.h>
+
+LibrariesInfoCache* LibrariesInfoCache::s_instance = nullptr;
+
+// This function is exposed by the ld_preload wrapper library.
+// This allows us to know if we have to reload the cache or not
+extern "C" unsigned long long dd_nb_calls_to_dlopen_dlclose() __attribute__((weak));
+
+LibrariesInfoCache::LibrariesInfoCache()
+{
+    LibrariesInfo.reserve(100);
+    NbCallsToDlopenDlclose = 0;
+    s_instance = this;
+    unw_set_iterate_phdr_function(unw_local_addr_space, LibrariesInfoCache::DlIteratePhdr);
+}
+
+LibrariesInfoCache* LibrariesInfoCache::Get()
+{
+    static LibrariesInfoCache Instance;
+    return &Instance;
+}
+
+LibrariesInfoCache::~LibrariesInfoCache()
+{
+    unw_set_iterate_phdr_function(unw_local_addr_space, dl_iterate_phdr);
+    s_instance = nullptr;
+}
+
+struct IterationData
+{
+public:
+    std::size_t Index;
+    LibrariesInfoCache* Cache;
+};
+
+void LibrariesInfoCache::UpdateCache()
+{
+    std::unique_lock l(_cacheLock);
+
+    auto nbCallsToDlopenDlclose = __builtin_expect(dd_nb_calls_to_dlopen_dlclose != nullptr, true) ? dd_nb_calls_to_dlopen_dlclose() : NbCallsToDlopenDlclose;
```

Comment:
To be sure I understand, the fallback case is when running without LD_PRELOAD? Maybe we should always resolve the modules in that case, instead of never?

Reply:
👍